### ISSUE TITLE
False positives when guarding map.get(key) with an expression

### ISSUE THREAD

**Issue by Edarke** [uber/NullAway#881]
Currently in v0.10.18, it seems that `map.get(key)` is recognized as non-null only if it is contained exactly in `if (map.containsKey(key))`. It would be great if this could be expanded to recognize simple expressions that imply `map.containsKey(key)`, such as the result being saved to a boolean without intervening calls to `map.remove` 

Here's two examples under the mild assumption that `map.containsKey` does not modify the keyset of the map.

```java
import java.util.Map;

public class NullAway {

  public static String getFirst(Map<String, String> map, String key1, String key2) {
    if (map.containsKey(key1)) {
      return map.get(key1); // Expression correctly flagged as Nonnull
    } else if (map.containsKey(key2)) {
      return map.get(key2); // Expression correctly flagged as Nonnull
    }
    return "";
  }

  public static String getOnly(Map<String, String> map, String key1, String key2) {
    boolean hasKey1 = map.containsKey(key1);
    boolean hasKey2 = map.containsKey(key2);

    if (hasKey1 && hasKey2) {
      throw new IllegalStateException();
    }

    if (hasKey1) {
      return map.get(key1); // Expression incorrectly flagged as Nullable
    } else if (hasKey2) {
      return map.get(key2); // Expression incorrectly flagged as Nullable
    }
    return "";
  }

  public static String getOnlyV2(Map<String, String> map, String key1, String key2) {
    if (map.containsKey(key1) && !map.containsKey(key2)) {
      return map.get(key1); // Expression incorrectly flagged as Nullable
    } else if (map.containsKey(key2) && !map.containsKey(key1)) {
      return map.get(key2); // Expression incorrectly flagged as Nullable
    }
    return "";
  }
}
```

**Comment by msridhar** [comment]
Thanks for the report!
 
Your `getOnly` example is a known issue (though I don't see an open GitHub issue on it).  The challenge is that tracking correlations between boolean variables and nullability conditions could get expensive (the local dataflow analysis is already the most costly part of NullAway), and it's hard to know how far to go with reasoning about boolean expressions (like `hasKey1 && hasKey2` in your example).  I'll be honest that I think it's unlikely we'll look into supporting code like this anytime soon.

The `getOnlyV2` example surprises me; I would expect we could handle that one.  I'll try to investigate what is going wrong.

**Comment by msridhar** [comment]
@Edarke I wrote a test case corresponding to `getOnlyV2`:

https://github.com/msridhar/NullAway/commit/265a7b19db66f3ae9687eb1e4ba4b68f682e836a

I'm not seeing the behavior you describe; the test passes without reporting any warnings.  Any idea what I need to do to reproduce?

**Comment by msridhar** [comment]
I'm going to go ahead and close as I can't reproduce the one issue that I would expect we can handle.  If there are further repro instructions please feel free to re-open.
